Dependency resolver using poetry

andyi2it opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://www.kubeflow.org/docs/about/contributing/ and developer guide https://github.com/kserve/kserve/blob/master/docs/DEVELOPER_GUIDE.md
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
4. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #2404 #2108 

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevent result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A
- [ ] Test B

- Logs

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [x] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [x] Have you made corresponding changes to the documentation?

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note

```

andyi2it committed:
Intial process of moving dependency management to poetry

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

update test dependencies

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Add strict dependency versions

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Remove setup.py and requirements.txt in python sdk

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

andyi2it committed:
Add toml and lock file and remove setup.py for aiffairness

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Add toml and lock file and remove setup.py for aixexplainer

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Update readme of kserve sdk

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Add toml and lock file and remove setup.py for alibiexplainer

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Add packages in toml file for alibi and aix explainers

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Add toml and lock file and remove setup.py for artexplainer

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Add toml and lock file and remove setup.py for lgbserver

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

andyi2it committed:
Add toml and lock file and remove setup.py for paddleserver

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Add toml and lock file and remove setup.py for pmmlserver

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Add toml and lock file and remove setup.py for sklearn server

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Add toml and lock file and remove setup.py for xgbserver

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

update author name in aliibi explainer

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

loosen pinned dependencies to allow patch level updates

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Add toml and lock file for custom model

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Add toml and lock file and remove setup.py for custom transformer

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Update readme

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

andyi2it committed:
Bump joblib version to 1.0.0 in sklearn server

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Add constraint for numpy <1.24.0

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Update python constraint >=3.7 for aixexplainer

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Bump joblib version to 1.2 in alibi

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Add docker build and push rules in makefile for art explainer and aiffairness

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Add missing dependency pillow to art explainer

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Update python Dockerfiles to use poetry

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Bump kserve version to 0.10.0

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Update python unit test workflow

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Update python sdk publish workflow

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Update e2e test gh-actions

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

fix e2e test workflow

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

revert kserve version to 0.10.0rc1

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

update qpext e2e test script

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

fix sklearn workflow step name

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

update authors for kserve sdk

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

update custom transformer description

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

update helm test workflow to use poetry

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

add constraint fastapi <=0.88.0

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Add script for updating release version in toml files

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Add separate job for py 3.7 build unit test

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Add pytest-cov dependency

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

andyi2it committed:
Fix missing dependency and update lock file

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Fix alibi explainer dependency errors

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Update version of kserve in all packages

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Add missing triton client

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Loosen protobuf and numpy dependency and replace table logger with tabulate

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Bump pillow from 9.0.1 to 9.3.0 in /python/custom_transformer

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

andyi2it committed:
update release version to 0.10.1

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

upgrade poetry to 1.4.0

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

fix custom model

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

resolve conflicts with making storage as optional dependency

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

fix kserve sdk unit test

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

fix alibi explainer build failure

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

improve dockerfiles by introducing parameters

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Remove obsolete requirements and setup.py

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

fix python test workflow

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

fix aixexplainer build failure

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

fix paddle server test failure

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

upgrade torchvision to 0.15.1 in custom model

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

Prune explainer dangling explainer images to free up space

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

andyi2it committed:
Increase timeout for test-fast and test-slow

test-fast and test-slow were failing with timeout and the timeout has
been increased

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

yuzisun commented on .github/workflows/python-test.yml:
```diff
@@ -28,31 +28,291 @@ jobs:
+          cd python
+          source alibiexplainer/.venv/bin/activate
           pytest --cov=alibiexplainer ./alibiexplainer
+  build-3-7:
```
Is there anything special handling for python 3.7 ? though this is going to be EOL soon in 6/2023
https://devguide.python.org/versions/

andyi2it replied:
Some of them, had requirements set as above 3.7 in their setup.py, though they were running in a matrix of 3.7, 3.8 and 3.9. So, it was segregated to make it explicit.
As you said, we can remove this once we remove support for 3.7, in the near future.

yuzisun commented on python/aiffairness/pyproject.toml:
```diff
@@ -0,0 +1,31 @@
+[tool.poetry]
+name = "aifserver"
+version = "0.10.1"
```
How do we update this version is the CI?

andyi2it replied:
Please refer, `update_release_version.sh`

yuzisun commented on python/alibiexplainer/pyproject.toml:
```diff
@@ -0,0 +1,31 @@
+pytest = "^7.2.0"
+pytest-cov = "^4.0.0"
+mypy = "^0.991"
+sklearnserver = { path = "../sklearnserver", develop = true }
```
Do we need sklearnserver here?

andyi2it replied:
They have used in test here, https://github.com/kserve/kserve/blob/master/python/alibiexplainer/tests/test_anchor_tabular.py#L36

yuzisun commented on python/artexplainer/pyproject.toml:
```diff
@@ -0,0 +1,30 @@
+[tool.poetry.dependencies]
+python = ">=3.9,<3.11"
+kserve = {path = "../kserve", develop = true}
+numpy = "^1.8.2"
```
this can be removed as kserve already depends on numpy

andyi2it replied:
Sure, will do.

yuzisun commented on hack/python-sdk/update_release_version.sh:
```diff
@@ -0,0 +1,46 @@
+  "python/xgbserver"
+)
+
+hack/python-sdk/update_release_version_helper.py "${VERSION}"
```
Do we need to manually run this script when releasing kserve?

yuzisun replied:
Created the issue to track this https://github.com/kserve/kserve/issues/2800

andyi2it committed:
Remove numpy in art explainer

Remove numpy dependency in art explainer as kserve already specifies the dependency

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

andyi2it committed:
Rebase 'Implement v2/open inference endpoints for OOTB python model servers' changes

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

andyi2it committed:
Update pyproject and poetry.lock for pandas and fastapi changes

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

andyi2it committed:
Added poetry version with poetry install

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

andyi2it committed:
Fix storage initializer

Remove alibi test for python 3.7

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

yuzisun commented on python/alibiexplainer.Dockerfile:
```diff
@@ -1,15 +1,47 @@
-ARG BASE_IMAGE=python:3.9-slim-bullseye
-FROM $BASE_IMAGE
+ARG PYTHON_VERSION=3.9
+ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim
```
```suggestion
ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim-bullseye
```

yuzisun commented on python/artexplainer.Dockerfile:
```diff
@@ -1,14 +1,45 @@
-ARG BASE_IMAGE=python:3.9-slim-bullseye
-FROM $BASE_IMAGE
+ARG PYTHON_VERSION=3.9
+ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim
```
```suggestion
ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim-bullseye
```

yuzisun commented on python/custom_model.Dockerfile:
```diff
@@ -1,15 +1,46 @@
-ARG BASE_IMAGE=python:3.9-slim-bullseye
-FROM $BASE_IMAGE
+ARG PYTHON_VERSION=3.9
+ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim
```
```suggestion
ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim-bullseye
```

yuzisun commented on python/custom_model_grpc.Dockerfile:
```diff
@@ -1,13 +1,45 @@
-ARG BASE_IMAGE=python:3.9-slim-bullseye
-FROM $BASE_IMAGE
+ARG PYTHON_VERSION=3.9
+ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim
```
```suggestion
ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim-bullseye
```

yuzisun commented on python/custom_transformer.Dockerfile:
```diff
@@ -1,15 +1,47 @@
-ARG BASE_IMAGE=python:3.9-slim-bullseye
-FROM $BASE_IMAGE
+ARG PYTHON_VERSION=3.9
+ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim
```
```suggestion
ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim-bullseye
```

yuzisun commented on python/custom_transformer_grpc.Dockerfile:
```diff
@@ -1,15 +1,48 @@
-ARG BASE_IMAGE=python:3.9-slim-bullseye
-FROM $BASE_IMAGE
+ARG PYTHON_VERSION=3.9
+ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim
```
```suggestion
ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim-bullseye
```

yuzisun commented on python/lgb.Dockerfile:
```diff
@@ -1,19 +1,50 @@
-ARG BASE_IMAGE=python:3.9-slim-bullseye
-FROM $BASE_IMAGE
+ARG PYTHON_VERSION=3.9
+ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim
```
```suggestion
ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim-bullseye
```

yuzisun commented on python/paddle.Dockerfile:
```diff
@@ -1,19 +1,50 @@
-ARG BASE_IMAGE=python:3.9-slim-bullseye
-FROM $BASE_IMAGE
+ARG PYTHON_VERSION=3.9
+ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim
```
```suggestion
ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim-bullseye
```

yuzisun commented on python/sklearn.Dockerfile:
```diff
@@ -1,14 +1,45 @@
-ARG BASE_IMAGE=python:3.9-slim-bullseye
-FROM $BASE_IMAGE
+ARG PYTHON_VERSION=3.9
+ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim
```
```suggestion
ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim-bullseye
```

yuzisun commented on python/storage-initializer.Dockerfile:
```diff
@@ -1,13 +1,28 @@
-ARG BASE_IMAGE=python:3.9-slim-bullseye
-FROM $BASE_IMAGE
+ARG PYTHON_VERSION=3.9
+ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim
```
```suggestion
ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim-bullseye
```

yuzisun commented on python/xgb.Dockerfile:
```diff
@@ -1,18 +1,52 @@
-ARG BASE_IMAGE=python:3.9-slim-bullseye
-FROM $BASE_IMAGE
+ARG PYTHON_VERSION=3.9
+ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim
```
```suggestion
ARG BASE_IMAGE=python:${PYTHON_VERSION}-slim-bullseye
```

andyi2it committed:
Updated python base image

Signed-off-by: Andrews Arokiam <andrews.arokiam@ideas2it.com>

yuzisun commented:
Awesome work!

/lgtm
/approve
